solo-io/gloo
# Bump solo-kit to pickup memory leak patches

### Change
Drastically slows memory leaks in Gloo and Discovery (evicted in weeks/months? rather than days) in active kubernetes environments. 

### Testing
1. Install gloo to GKE, `glooctl install gateway enterprise --license-key $LICENSE_KEY -f https://storage.googleapis.com/gloo-ee-helm/charts/gloo-ee-1.3.0-beta6.tgz`
2. Run script to generate 50 namespaces with 50 services each:
```shell script
for i in $(seq 50); do
kubectl create ns dev-$i
kubectl create deployment nothing --image=rancher/pause:3.1 -n dev-$i
    for j in $(seq 50); do
    kubectl -n dev-$i expose deployment nothing --name nothing-$j --port 8080
    done
done
```
3. Install petstore so we can route to it (likely optional), `kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo/v1.2.9/example/petstore/petstore.yaml`
4. Add a route to petstore (likely optional), `glooctl add route --name animal\
   --path-exact /animals \
   --dest-name default-petstore-8080 \
   --prefix-rewrite /api/pets`
5. Build images so we can get symbols when profiling Gloo components, `TAGGED_VERSION=vkdorosh GCLOUD_PROJECT_ID=solo-public make docker docker-push -B`
6. Update deployments to use new images, `kubectl edit deployment/discovery -n gloo-system` (and gloo) to use `kdorosh` image
7. Monitor memory use over time, `kubectl top pod -n gloo-system --containers` confirm memory leak over time
8. Port forward so we can profile our containers, `kubectl port-forward -n gloo-system deployment/discovery 9091`
9. Profile with interactive UI, `pprof -http=:3000 _output/discovery-discovery-amd64 http://localhost:9091/debug/pprof/heap`

Profiled and monitored memory usage/object allocation over time to ensure that the memory leak is slowed. In steady state at large scale (with 50 namespaces x 50 services each) we leak memory much slower until Gloo & Discovery get evicted. I have ideas to fix the root cause for good, but they require more wide-sweeping changes, and this is meant to be a patch to reduce pain until the rest of the work is done.
